RecoVertex/PrimaryVertexProducer/plugins/PrimaryVertexProducer.cc

```diff
@@ -149,6 +155,26 @@ void PrimaryVertexProducer::produce(edm::Event& iEvent, const edm::EventSetup& i
     edm::LogError("UnusableBeamSpot") << "Beamspot with invalid errors " << beamVertexState.error().matrix();
   }
 
+  //if this is a recovery iteration, check if we already have a valid PV
+  if (isRecoveryIteration) {
+    edm::Handle<reco::VertexCollection> oldPVs;
+    iEvent.getByToken(recoveryVtxToken, oldPVs);
+    const reco::VertexCollection* oldVertices;
+    oldVertices = oldPVs.product();
+    //look for the first valid (not-BeamSpot) vertex
+    for (size_t i = 0; i < oldVertices->size(); ++i) {
+      if (!((*oldVertices)[i].isFake())) {
+        //found a valid vertex, write the first one to the collection and return
+        //otherwise continue with regular vertexing procedure
+        auto result = std::make_unique<reco::VertexCollection>();
+        reco::VertexCollection& vColl = (*result);
+        vColl.push_back((*oldVertices)[i]);
```

Comment:
```suggestion
    for (auto const& old :  oldVertices) {
      if (!(old.isFake())) {
        //found a valid vertex, write the first one to the collection and return
        //otherwise continue with regular vertexing procedure
        auto result = std::make_unique<reco::VertexCollection>();
        result->push_back(old);
```